hypothesis-python/src/hypothesis/strategies/_internal/strategies.py

```diff
@@ -481,14 +484,34 @@ def validate(self) -> None:
         """
         if self.validate_called:
             return
-        try:
-            self.validate_called = True
-            self.do_validate()
-            self.is_empty
-            self.has_reusable_values
-        except Exception:
-            self.validate_called = False
-            raise
+        # we need to set validate_called before calling do_validate, for
+        # recursive / deferred strategies. But if a thread switches after
+        # validate_called but before do_validate, we might have a strategy
+        # which does weird things like drawing when do_validate would error but
+        # its params are technically valid (e.g. a param was passed as 1.0
+        # instead of 1) and get into weird internal states.
+        #
+        # To fix this we track a lock per strategy.
+        #
+        # We also maintain a single lock for creating these locks, because we
+        # don't want two threads to create two different locks which they both
+        # think is canonical for their strategy.
+        with validate_locks_lock:
+            try:
+                validate_lock = validate_locks[self]
+            except KeyError:
+                validate_lock = Lock()
+                validate_locks[self] = validate_lock
+
+        with validate_lock:
+            try:
+                self.validate_called = True
```

Comment:
Should `validate_called = True` be moved below `do_validate()`, given the early return outside lock on l.486?

That would leave open the possibility of `do_validate` called multiple time in an initialization race (unless explicitly checked inside), probably harmless though?

Reply:
Hm... per comment, this isn't sufficient. Maybe the early return needs to move inside critical region, or if that is too expensive possibly a two-stage process, i.e. early-return like
```
if self.validate_called and not self.validate_in_progress:
    return
```

I'm not thinking clearly right now, so please consider this a hint not a recipe

Reply:
`if self.validate_called and not self.validate_in_progress:` is also going to run into an infinite recursion, but possibly what we could do is set `validate_calleds: dict[int, bool]` which tracks `thread_id: validate_called`, and only return if vlidate has been called on this thread before. This would be lock-free and allow for concurrent validates. If `threading.get_ident()` is not expensive then this could work. I'll test it

Reply:
The downside of this is that all threads rerun validation for all strategies. It's a tradeoff between singlethreaded and multithreaded performance. I'm defaulting to prioritizing singlethreaded performance, but I could see us changing this in the future (or now, if people have preferences)

Reply:
I'd prefer to keep prioritizing single-threaded perf for now, at least when it's not super-lopsided.

Potential caveat in this case: is it safe to concurrently run validation for a strategy in multiple threads?  If not, we should probably go with the smallish cost of a lock.

Reply:
concurrent validation should be safe, yeah. (in theory, and I've tested a full run in practice)